packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +51,400 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
     assert.object(arg.schema, "schema static");
     const schema = normalizeObjectSchema(arg.schema as ObjectSchema);
-    schema.constructor = arg;
+    schema.ctor = arg;
     return schema;
-  } else {
-    // TODO: Determine if we still want to support this
-    if (Array.isArray(arg.properties)) {
-      if (flags.ALLOW_VALUES_ARRAYS) {
-        return normalizeObjectSchema({
-          ...arg,
-          properties: Object.fromEntries(arg.properties.map(({ name, ...rest }) => [name, rest])),
-        });
-      } else {
-        throw new Error("Array of properties are no longer supported");
-      }
+  }
+
+  // ---- THIS IF BLOCK HAS NOT YET BEEN REWRITTEN ----
+  // TODO: Determine if we still want to support this (should show warning to users of future deprecation)
+  if (Array.isArray(arg.properties)) {
+    if (flags.ALLOW_VALUES_ARRAYS) {
+      return normalizeObjectSchema({
+        ...arg,
+        // Build the PropertiesTypes object
+        properties: Object.fromEntries(arg.properties.map(({ name, ...rest }) => [name, rest])),
+      });
     }
-    return {
-      constructor: undefined,
-      name: arg.name,
-      primaryKey: arg.primaryKey,
-      asymmetric: arg.asymmetric || false,
-      embedded: arg.embedded || false,
-      properties: Object.fromEntries(
-        Object.entries(arg.properties).map(([name, property]) => {
-          const canonicalPropertySchema = normalizePropertySchema(name, property);
-          // A primary key is always indexed
-          if (name === arg.primaryKey) {
-            canonicalPropertySchema.indexed = true;
-          }
-          validateCanonicalPropertySchema(arg.name, canonicalPropertySchema);
-          return [name, canonicalPropertySchema];
-        }),
-      ),
-    };
+    throw new Error("Array of properties are no longer supported. Use an object instead.");
   }
+  // --------------------------------------------------
+
+  return {
+    name: arg.name,
+    primaryKey: arg.primaryKey,
+    asymmetric: !!arg.asymmetric,
+    embedded: !!arg.embedded,
+    properties: normalizePropertySchemas(arg.name, arg.properties, arg.primaryKey),
+  };
+}
+
+/**
+ * Transform user-provided property schemas into their canonical forms.
+ */
+function normalizePropertySchemas(
+  objectName: string,
+  schemas: PropertiesTypes,
+  primaryKey?: string,
+): Record<string, CanonicalObjectSchemaProperty> {
+  const normalizedSchemas: Record<string, CanonicalObjectSchemaProperty> = {};
+  for (const propertyName in schemas) {
+    normalizedSchemas[propertyName] = normalizePropertySchema(
+      { objectName, propertyName },
+      schemas[propertyName],
+      primaryKey === propertyName,
+    );
+  }
+
+  return normalizedSchemas;
 }
 
-// TODO: Extend to handle all short hands
+/**
+ * Transform a user-provided property schema into its canonical form.
+ */
 export function normalizePropertySchema(
-  name: string,
+  name: ObjectAndPropertyName,
   schema: string | ObjectSchemaProperty,
```

Comment:
I would at first also interpret `schema` as an object's schema. Although, that word is currently used for many things (in the existing code base). A Realm's schema (e.g. when creating a Realm instance) refers to the Realm and means an array of object schemas (`CanonicalRealmSchema`). An object schema refers to the object (`CanonicalObjectSchema`), and a property schema refers to the object's property (`CanonicalObjectSchemaProperty` or if renamed `PropertySchema`).

So in this sense, the parameter `schema` would mean whatever it means in the context. Given the already-existing concepts of a schema, having `schema` only mean an object's schema may in itself be confusing.

Since `type` is also used in this part of the code to mean the actual `type`field, it may be interpreted differently as well.

Replacing `ObjectAndPropertyName` seems like a very good idea! I'll try to see how I can narrow down from the `string | ObjectSchemaProperty` when needed.

Reply:
If we can converge on some terminology, i'm happy with `schema` being `ObjectSchema` or `ModelSchema` which contains a list of properties that are defined by a `PropertySchema`.  In which case, if the property is defined by an object, this would be of type `PropertySchema`, and if it's defined by a string, then it would be of type `PropertySchemaShorthand` which is just an alias for `string`.

With this terminology and typing, any variable referring to a `schema` should be prefixed as `const objectSchema` or `const propertySchema`. Variable referring to the shorthand would be `const propertySchemaShorthand` or `const propertyShorthand`.  Bit wordy, but I think this is more clear.

Reply:
Yeah it makes it clearer to distinguish them a bit more. Btw is this what you mean?

Having:
```ts
// This is the object of properties
export type PropertiesTypes = {
  [keys: string]: PropertySchema | PropertySchemaShorthand;
};
```

Reply:
Almost.  This is what I mean.

```typescript
export type ObjectSchema = { 
  name: string;
  primaryKey?: string;
  embedded?: boolean;
  asymmetric?: boolean;
  properties: PropertiesTypes;
};

export type PropertySchema = { // <-- copied from CanonicalObjectSchema
  name: string;
  type: PropertyTypeName;
  optional: boolean;
  indexed: boolean;
  mapTo: string; // TODO: Make this optional and leave it out when it equals the name
  objectType?: string;
  property?: string;
  default?: unknown;
};

export type PropertiesTypes = {
  [key: string]: PropertySchema | PropertySchemaShorthand;
};
```

Reply:
Gotcha 👍  I'll leave `name` out though, it's only for the canonical form.